nvflare/app_opt/xgboost/tree_based/shareable_generator.py

```diff
@@ -24,20 +24,37 @@
     if not prev_model:
-        model_update["learner"]["gradient_booster"]["model"]["gbtree_model_param"]["num_parallel_tree"] = "1"
         return model_update
     else:
-        # Always 1 tree, so [0]
-        best_iteration = int(prev_model["learner"]["attributes"]["best_iteration"])
-        best_ntree_limit = int(prev_model["learner"]["attributes"]["best_ntree_limit"])
-        num_trees = int(prev_model["learner"]["gradient_booster"]["model"]["gbtree_model_param"]["num_trees"])
-        prev_model["learner"]["attributes"]["best_iteration"] = str(best_iteration + 1)
-        prev_model["learner"]["attributes"]["best_ntree_limit"] = str(best_ntree_limit + 1)
-        prev_model["learner"]["gradient_booster"]["model"]["gbtree_model_param"]["num_trees"] = str(num_trees + 1)
-        append_info = model_update["learner"]["gradient_booster"]["model"]["trees"][0]
-        append_info["id"] = num_trees
-        prev_model["learner"]["gradient_booster"]["model"]["trees"].append(append_info)
-        prev_model["learner"]["gradient_booster"]["model"]["tree_info"].append(0)
+        # Append all trees
+        pre_num_parallel_tree = int(
+            prev_model["learner"]["gradient_booster"]["model"]["gbtree_model_param"]["num_parallel_tree"]
+        )
+        pre_best_iteration = int(prev_model["learner"]["attributes"]["best_iteration"])
+        pre_best_ntree_limit = int(prev_model["learner"]["attributes"]["best_ntree_limit"])
+        pre_num_trees = int(prev_model["learner"]["gradient_booster"]["model"]["gbtree_model_param"]["num_trees"])
+
+        add_num_parallel_tree = int(
+            model_update["learner"]["gradient_booster"]["model"]["gbtree_model_param"]["num_parallel_tree"]
+        )
+        add_best_iteration = int(model_update["learner"]["attributes"]["best_iteration"])
+        add_best_ntree_limit = int(model_update["learner"]["attributes"]["best_ntree_limit"])
+        add_num_trees = int(model_update["learner"]["gradient_booster"]["model"]["gbtree_model_param"]["num_trees"])
+
+        prev_model["learner"]["gradient_booster"]["model"]["gbtree_model_param"]["num_parallel_tree"] = str(
+            pre_num_parallel_tree + add_num_parallel_tree
+        )
+        prev_model["learner"]["attributes"]["best_iteration"] = str(pre_best_iteration + add_best_iteration)
+        prev_model["learner"]["attributes"]["best_ntree_limit"] = str(pre_best_ntree_limit + add_best_ntree_limit)
+        prev_model["learner"]["gradient_booster"]["model"]["gbtree_model_param"]["num_trees"] = str(
+            pre_num_trees + add_num_trees
+        )
+
+        append_info = model_update["learner"]["gradient_booster"]["model"]["trees"]
+        for tree_ct in range(add_num_trees):
+            append_info[tree_ct]["id"] = pre_num_trees + tree_ct
+            prev_model["learner"]["gradient_booster"]["model"]["trees"].append(append_info[tree_ct])
+            prev_model["learner"]["gradient_booster"]["model"]["tree_info"].append(0)
```

Comment:
what is this "tree_info" for?

Reply:
part of xgboost model (https://xgboost.readthedocs.io/en/latest/tutorials/saving_model.html), defining class information (check this issue https://github.com/dmlc/xgboost/issues/6623)